src/pages/the-elements-of-ui-engineering/index.pt-br.md

```diff
@@ -0,0 +1,61 @@
+
+Acredito fortemente que você pode "começar em qualquer lugar" e não necessite de aprender tecnologias em uma ordem específica. Mas também dou muito valor em ganhar perícia em algo. Pessoalmente me interesso mais na criação de interfaces de usuário.
+
+**Tenho refletido sobre o que *sei* sobre e que considero valioso.** Claro, estou familiarizado com algumas tecnologias (por exemplo JavaScript e React). Mas as lições mais importantes da experiência são elusivas. Nunca tentei colocá-las em palavras. Essa é minha primeira tentativa de listar e descrever algumas delas.
+
+---
+
+Existem diversas "trilhas de aprendizagem" sobre tecnologias e bibliotecas. Qual biblioteca estará em destaque em 2019? E em 2020? Devo aprender Vue ou React? Angular? E sobre Redux ou Rx? Devo aprender Apollo? REST ou GraphQL? É fácil se perder. E se o autor estiver errado?
+
+**Meus maiores avanços de aprendizado não foram sobre uma tecnologia específica.** Ao invés disso, eu aprendia bem mais quando tinha dificuldades em resolver algum problema específico de UI. Às vezes, mais tarde eu descobria bibliotecas ou padrões que me ajudavam. Em outros casos, eu criava minhas próprias soluções (boas e ruins).
+
+É essa combinação de entender os *problemas*, experimentar *soluções* e aplicar diferentes *estratégias* que me levaram às experiências de aprendizado mais recompensadoras. **Esse artigo foca apenas nos problemas.**
+
+---
+
+Se você já trabalhou com interfaces de usuário, provavelmente já teve de lidar com alguns desses desafios - seja diretamente ou usando uma biblioteca. Em ambos os casos, incentivo você a criar uma pequena aplicação _sem_ bibliotecas e brincar reproduzindo e resolvendo esses problemas. Não há uma solução correta para nenhum deles. O aprendizado vem de explorar o escopo do problema e tentar diferentes possíveis soluções.
+
+---
+
+* **Consistência.** Você clica em um botão "Curtir" e o texto atualiza para: "Você e mais 3 amigos curtiram esse post." Você clica nele novamente e o texto retorna ao anterior. Parece fácil. Mas talvez um label assim está presente em diversos locais na tela. Talvez tenha uma outra indicação visual (como o fundo do botão) que deve ser alterado. A lista de "curtidores" que foi anteriormente obtida do servidor e está visível quando se passa por cima agora deve incluir seu nome. Se você navegar para uma outra página e voltar, o post não deve "esquecer" que foi curtido. Até mesmo a consistência local *por si só* cria uma lista de desafios. Mas outros usuários podem também modificar os dados que estamos mostrando (por exemplo ao curtir um post que estamos vendo). Como mantemos os mesmos dados sincronizados em diferentes partes da tela? Como e quando fazemos os dados locais serem consistentes com o servidor e como fazemos o caminho contrário?
```

Comment:
Original: "How and when do we make the local data consistent with the server, and the other way around?"

```diff
+ Como e quando fazemos os dados locais serem consistentes com o servidor e vice-versa?
- Como e quando fazemos os dados locais serem consistentes com o servidor e como fazemos o caminho contrário?
```

```suggestion
* **Consistência.** Você clica em um botão "Curtir" e o texto atualiza para: "Você e mais 3 amigos curtiram esse post." Você clica nele novamente e o texto retorna ao anterior. Parece fácil. Mas talvez um label assim está presente em diversos locais na tela. Talvez tenha uma outra indicação visual (como o fundo do botão) que deve ser alterado. A lista de "curtidores" que foi anteriormente obtida do servidor e está visível quando se passa por cima agora deve incluir seu nome. Se você navegar para uma outra página e voltar, o post não deve "esquecer" que foi curtido. Até mesmo a consistência local *por si só* cria uma lista de desafios. Mas outros usuários podem também modificar os dados que estamos mostrando (por exemplo ao curtir um post que estamos vendo). Como mantemos os mesmos dados sincronizados em diferentes partes da tela? Como e quando fazemos os dados locais serem consistentes com o servidor e vice-versa?
```